executions/graphql-kotlin-transaction-batcher/src/main/kotlin/com/expediagroup/graphql/transactionbatcher/publisher/TriggeredPublisher.kt

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2022 Expedia, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.expediagroup.graphql.transactionbatcher.publisher
+
+import com.expediagroup.graphql.transactionbatcher.transaction.BatcheableTransaction
+import com.expediagroup.graphql.transactionbatcher.transaction.TransactionBatcherCache
+import org.reactivestreams.Publisher
+import org.reactivestreams.Subscriber
+import org.reactivestreams.Subscription
+import org.slf4j.LoggerFactory
+
+/**
+ * Interface representing a publisher with input [TInput] type and output [TOutput] type
+ */
+@Suppress(
+    "ReactiveStreamsSubscriberImplementation",
+    "UNCHECKED_CAST"
+)
+fun interface TriggeredPublisher<TInput, TOutput> {
+    /**
+     * Given an input of type [TInput] create a cold [Publisher] that will produce a [Publisher] of type [TOutput] of n elements
+     * that maps to the size of the input [List] of [TInput]
+     * order is important so make sure to produce elements in the same order of [input]
+     */
+    fun produce(input: List<TInput>): Publisher<TOutput>
+
+    /**
+     * Attempts to collect values from [cache] first and then [produce]
+     *
+     * Example:
+     * if [TriggeredPublisher] is of type <Int, Int> and [cache] resolves [1, null, 3, null, 5, 6]
+     * we will attempt to produce elements for index 1 and 3
+     * when [produce] stream completes we will complete futures from either values resolved from [cache] or from [produce]
+     */
+    fun trigger(
+        batcheableTransactions: List<BatcheableTransaction<TInput, TOutput>>,
+        cache: TransactionBatcherCache
+    ) {
+
+        val values = batcheableTransactions.map { batcheableTransaction ->
```

Comment:
I think we can simplify this to something like

```kotlin
val transactionsNotInCache = batchableTransactions.filterNot { transaction -> cache.get(transaction.key) != null }
```

*instead of doing the `cache.get` we could also add `cache.contains` which I think would clarify the intent better

Reply:
actually I'd also apply `map(BatcheableTransaction<TInput, TOutput>::input)` at the same time so we would just pass a list